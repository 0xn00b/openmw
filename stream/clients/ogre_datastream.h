#ifndef MANGLE_STREAM_OGRECLIENT_H
#define MANGLE_STREAM_OGRECLIENT_H

#include <OgreDataStream.h>
#include <assert.h>
#include "../stream.h"

namespace Mangle {
namespace Stream {

/** An OGRE DataStream that wraps a Mangle::Stream input.

    This has been built and tested against OGRE 1.6.2. You might have
    to make your own modifications if you're working with newer (or
    older) versions.
 */
class MangleDataStream : public Ogre::DataStream
{
  StreamPtr inp;

  void init()
    {
      // Get the size, if possible
      if(inp->hasSize)
        mSize = inp->size();
    }

 public:
  /// Constructor without name
  MangleDataStream(StreamPtr _inp)
    : inp(_inp) { init(); }

  /// Constructor for a named data stream
  MangleDataStream(const Ogre::String &name, StreamPtr _inp)
    : inp(_inp), Ogre::DataStream(name) { init(); }

  // Only implement the DataStream functions we have to implement

  size_t read(void *buf, size_t count)
    { return inp->read(buf,count); }

  void skip(long count)
    {
      assert(inp->isSeekable && inp->hasPosition);
      inp->seek(inp->tell() + count);
    }

  void seek(size_t pos)
    { assert(inp->isSeekable); inp->seek(pos); }

  size_t tell() const
    { assert(inp->hasPosition); return inp->tell(); }

  bool eof() const { return inp->eof(); }

  /// Does nothing
  void close() {}
};

}} // namespaces
#endif
