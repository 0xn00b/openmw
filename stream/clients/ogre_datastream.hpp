#ifndef MANGLE_STREAM_OGRECLIENT_H
#define MANGLE_STREAM_OGRECLIENT_H

#include <OgreDataStream.h>
#include <assert.h>
#include "../stream.hpp"

namespace Mangle {
namespace Stream {

/** An OGRE DataStream that wraps a Mangle::Stream input.

    This has been built and tested against OGRE 1.6.2. You might have
    to make your own modifications if you're working with newer (or
    older) versions.
 */
class Mangle2OgreStream : public Ogre::DataStream
{
  StreamPtr inp;

  void init()
    {
      // Get the size, if possible
      if(inp->hasSize)
        mSize = inp->size();
    }

 public:
  /// Constructor without name
  Mangle2OgreStream(StreamPtr _inp)
    : inp(_inp) { init(); }

  /// Constructor for a named data stream
  Mangle2OgreStream(const Ogre::String &name, StreamPtr _inp)
    : Ogre::DataStream(name), inp(_inp) { init(); }

  // Only implement the DataStream functions we have to implement

  size_t read(void *buf, size_t count)
    { return inp->read(buf,count); }

  void skip(long count)
    {
      assert(inp->isSeekable && inp->hasPosition);
      inp->seek(inp->tell() + count);
    }

  void seek(size_t pos)
    { assert(inp->isSeekable); inp->seek(pos); }

  size_t tell() const
    { assert(inp->hasPosition); return inp->tell(); }

  bool eof() const { return inp->eof(); }

  /// Does nothing
  void close() {}
};

}} // namespaces
#endif
