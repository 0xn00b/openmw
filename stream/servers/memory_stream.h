#ifndef MANGLE_STREAM_MEMSERVER_H
#define MANGLE_STREAM_MEMSERVER_H

#include <assert.h>
#include "../stream.h"

namespace Mangle {
namespace Stream {

/** A Stream wrapping a memory buffer

    This will create a fully seekable stream out any pointer/length
    pair you give it.
 */
class MemoryStream : public Stream
{
  const void *data;
  size_t length, pos;

 public:
 MemoryStream(const void *ptr, size_t len)
   : data(ptr), length(len), pos(0)
  {
    isSeekable = true;
    hasPosition = true;
    hasSize = true;
  }

  size_t read(void *buf, size_t count)
  {
    assert(pos <= length);

    // Don't read more than we have
    if(count > (length - pos))
      count = length - pos;

    // Copy data
    if(count)
      memcpy(buf, ((char*)data)+pos, count);

    // aaand remember to increase the count
    pos += count;

    return count;
  }

  void seek(size_t _pos)
  {
    pos = _pos;
    if(pos > length)
      pos = length;
  }

  size_t tell() const { return pos; }
  size_t size() const { return length; }
  bool eof() const { return pos == length; }
};

}} // namespaces
#endif
