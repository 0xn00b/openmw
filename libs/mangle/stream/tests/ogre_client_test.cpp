#include "../servers/memory_stream.hpp"
#include "../clients/ogre_datastream.hpp"
#include <iostream>

using namespace Mangle::Stream;
using namespace Ogre;
using namespace std;

int main()
{
  StreamPtr inp(new MemoryStream("hello world", 11));
  DataStreamPtr p(new Mangle2OgreStream("hello", inp));
  cout << "Name: " << p->getName() << endl;
  cout << "As string: " << p->getAsString() << endl;
  cout << "pos=" << p->tell() << " eof=" << p->eof() << endl;
  p->seek(0);
  cout << "pos=" << p->tell() << " eof=" << p->eof() << endl;
  p->skip(5);
  p->skip(-2);
  cout << "pos=" << p->tell() << " eof=" << p->eof() << endl;
  return 0;
}
