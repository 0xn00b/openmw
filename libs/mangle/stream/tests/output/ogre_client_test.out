Name: hello
As string: hello world
pos=11 eof=1
pos=0 eof=0
pos=3 eof=0
